debezium/dbz#2121 Blog post about recent JDBC sink support in Debezium Server (#1277)
debezium/debezium.github.io: _posts/2026-06-22-kafka-less-migration.adoc

While this was true in the project's early days, it is no longer the case, as we explained in detail,
together with several other common misconceptions, in our recent article, https://debezium.io/blog/2026/05/22/what-nobody-explains-about-debezium-2026/["What Nobody Explains About Debezium in 2026 (But Should)"].

In this article, we’ll build on that discussion by showing the recently added JDBC sink support in Debezium Server (DS) and the Debezium Management Platform (DMP),
which opens up Kafka-less data replication and migration scenarios.

+++<!-- more -->+++

Data integration, whether for one-shot migrations or continuous replication between databases, is one of the core use cases Debezium supports.
Debezium has always focused on the CDC side of the problem: capturing changes from a database and turning them into a stream of change events.
In the past, users had to develop their own consumer to write those events into a target database.
To address this, we introduced the https://debezium.io/documentation/reference/stable/connectors/jdbc.html[JDBC Sink Connector] in the `2.2.0.Beta1` release in 2023,
enabling end-to-end replication out of the box.

> Today, the same capability is available through the https://debezium.io/documentation/reference/stable/operations/debezium-server.html#_jdbc[DS JDBC sink] and within the https://debezium.io/documentation/reference/stable/operations/debezium-platform.html[DMP].

Reviewer: I would put the link only on DS instead of DS JDBC sink since we are pointing to the debezium server page
```suggestion
Today, the same capability is available through the https://debezium.io/documentation/reference/stable/operations/debezium-server.html#_jdbc[DS] JDBC sink and within the https://debezium.io/documentation/reference/stable/operations/debezium-platform.html[DMP].
```

Author: Effectively the link point directly to the jdbc sink (`debezium-server.html#_jdbc`) this is the reason of its name. WDYT?

Reviewer: Weird because if I open link I see
<img width="1678" height="926" alt="image" src="https://github.com/user-attachments/assets/cee11249-7d85-45d7-b59a-b5258690d688" />

Author: Yes because the link is the `stable` that currently points to the last 3.5 release. It will work once the `3.6.0.Final` will be released like https://debezium.io/documentation/reference/3.6/operations/debezium-server.html#_jdbc

"INSERT INTO inventory.customers (first_name, last_name, email) VALUES ('John', 'Doe', 'john.doe@example.com');"
----

and verify that it is replicated on the destination

[source, shell]
----
docker exec -it mysql mysql -uroot -pdebezium target_db -e "SELECT * FROM migration_inventory_customers;"
----

There we go! You have now a data replication pipeline migrating data from PostgreSQL to MySQL, in real time, just running a single container!

[source, shell]
----
+------+------------+-----------+-----------------------+
| id   | first_name | last_name | email                 |
+------+------------+-----------+-----------------------+
| 1001 | Sally      | Thomas    | sally.thomas@acme.com |
| 1002 | George     | Bailey    | gbailey@foobar.com    |
| 1003 | Edward     | Walker    | ed@walker.com         |
| 1004 | Anne       | Kretchmar | annek@noanswer.org    |
| 1005 | John       | Doe       | john.doe@example.com  |
+------+------------+-----------+-----------------------+
----

[NOTE]
====
I strongly suggest to keep an eye on https://github.com/debezium/dbz/issues/1731[DBZ#1731-Migration of Debezium Server to the Debezium Quarkus Extensions],

> that will permit to have a native build of DS sever so you could have the same data migration pipeline with just a single native executable.

Reviewer: Maybe I would add this as one of the reason to choose DS here https://github.com/debezium/debezium.github.io/pull/1277/changes#diff-5069cb8a072408b1c6a193ad0084820de12916c4e5e37e7c3940d7373e95bca1R26

Author: Since is something not yet ready I preferred to keep those things in a side note. I would like to know also what are the others though on this.

Reviewer: In my opinion the native build is one of the main benefits of choosing DS. since it supports it, unlike Kafka Connect.

Author: Added. Let me know if sounds better now.

</div>
++++

Then create the source and the destination

*Create the source*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-source.png" style="max-width:100%;" class="responsive-image" alt="Create the source">
</div>
++++

*Create the destination*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-destination.png" style="max-width:100%;" class="responsive-image" alt="Create the destination">
</div>
++++

Once you have all the pieces ready, you can design the pipeline that will replicate data from the source to the destination database

*Design the pipeline*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/design-pipeline.png" style="max-width:100%;" class="responsive-image" alt="Pipeline design">
</div>
++++

and the finally create it

*Create the pipeline*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-pipeline.png" style="max-width:100%;" class="responsive-image" alt="Create pipeline">
</div>
++++

> It is worth to mention that you can do the same thing via the REST API with the following

Reviewer: Instead of listing all of them below, don't we have a link to the REST API documentation?

Author: Unfortunately no. The API documentation is only available on a live instance. 
This is a good point, we should generate it at build time and publish together with the doc.

-H 'Content-Type: application/json' \
-d '{
"name": "jdbc-destination",
"type": "jdbc",
"schema": "dummy",
"connection": { "id": 2 },
"config": {
"insert.mode": "upsert",
"primary.key.mode": "record_key",
"schema.evolution": "basic"
}
}'
----

.Create pipeline (use the source and destination ids from steps 3 and 4):
[source, shell]
----
curl -X POST http://platform.debezium.io/api/pipelines \
-H 'Content-Type: application/json' \
-d '{
"name": "database-migration",
"source": { "id": 1 },
"destination": { "id": 1 },
"logLevel": "INFO"
}'
----

== Conclusion
We have seen how simple it is to build a data migration pipeline without requiring a Kafka cluster.

There are still some limitations in the current DS JDBC sink implementation.
For example, while Kafka Connect (KC) allows processing to be parallelized by configuring `tasks.max`, this capability is not yet available in DS.

Similarly, if you want to process records from only a specific set of tables, there is no direct equivalent to the KC topics configuration.
Instead, you need to filter records at the source, using the `table.include.list` configuration, or apply a transformation to select the desired events.

> These are known limitations, and addressing them is already on our radar for future releases.

Reviewer: Perhaps we could also mention here that we are ready to discuss any specific requirements or requests that the community may have for DS and the DMP, in order to align our engagement with the community's priorities.

---
layout: post
title:  "Building Kafka-Less Data Integration Pipelines with Debezium"
date:   2026-06-17
tags: [debezium-server, debezium-platform, replication, migration]
author: mfvitale
---

> There is a recurring narrative that Debezium requires Kafka.

Reviewer 2: I don't know if it's the case to underline this narrative even if it's a good introduction for the topic

Reviewer 3: Agreed. I think we can frame it more like "Debezium is often associated with Kafka, but did you know ...". I think it lands the intent but in a more subtle way.

While this was true in the project's early days, it is no longer the case, as we explained in detail,
together with several other common misconceptions, in our recent article, https://debezium.io/blog/2026/05/22/what-nobody-explains-about-debezium-2026/["What Nobody Explains About Debezium in 2026 (But Should)"].

> In this article, we’ll build on that discussion by showing the recently added JDBC sink support in Debezium Server (DS) and the Debezium Management Platform (DMP),

Reviewer 4: Maybe add links to DS and DPM also here where we mention it for the first time.

which opens up Kafka-less data replication and migration scenarios.

+++<!-- more -->+++

Data integration, whether for one-shot migrations or continuous replication between databases, is one of the core use cases Debezium supports.
Debezium has always focused on the CDC side of the problem: capturing changes from a database and turning them into a stream of change events.
In the past, users had to develop their own consumer to write those events into a target database.
To address this, we introduced the https://debezium.io/documentation/reference/stable/connectors/jdbc.html[JDBC Sink Connector] in the `2.2.0.Beta1` release in 2023,
enabling end-to-end replication out of the box.

Today, the same capability is available through the https://debezium.io/documentation/reference/stable/operations/debezium-server.html#_jdbc[DS JDBC sink] and within the https://debezium.io/documentation/reference/stable/operations/debezium-platform.html[DMP].
This is not intended to replace Kafka Connect (KC) connectors.

> Instead, it provides an alternative for users who do not have a KC cluster, or who do not need the scalability and high availability that KC offers,

Reviewer 4: ```suggestion
Instead, it provides an alternative for the users who do not have a KC cluster, or who do not need the scalability and high availability that KC offers,
```

|        109 |        5 |
+------------+----------+
----
Then let’s add a new customer.

[source, shell]
----
docker exec -it postgres psql -U postgres -d postgres -c \
"INSERT INTO inventory.customers (first_name, last_name, email) VALUES ('John', 'Doe', 'john.doe@example.com');"
----

and verify that it is replicated on the destination

[source, shell]
----
docker exec -it mysql mysql -uroot -pdebezium target_db -e "SELECT * FROM migration_inventory_customers;"
----

There we go! You have now a data replication pipeline migrating data from PostgreSQL to MySQL, in real time, just running a single container!

[source, shell]
----
+------+------------+-----------+-----------------------+
| id   | first_name | last_name | email                 |
+------+------------+-----------+-----------------------+
| 1001 | Sally      | Thomas    | sally.thomas@acme.com |
| 1002 | George     | Bailey    | gbailey@foobar.com    |
| 1003 | Edward     | Walker    | ed@walker.com         |
| 1004 | Anne       | Kretchmar | annek@noanswer.org    |
| 1005 | John       | Doe       | john.doe@example.com  |
+------+------------+-----------+-----------------------+
----

[NOTE]
====

> I strongly suggest to keep an eye on https://github.com/debezium/dbz/issues/1731[DBZ#1731-Migration of Debezium Server to the Debezium Quarkus Extensions]

Reviewer 3: ```suggestion
We recommend keeping an eye on https://github.com/debezium/dbz/issues/1731[DBZ#1731-Migration of Debezium Server to the Debezium Quarkus Extensions]
```

and https://github.com/debezium/dbz/issues/2107[DBZ#2108-Debezium Server Native],

> that will permit to have a native build of DS sever so you could have the same data migration pipeline with just a single native executable.

Reviewer 3: ```suggestion
that will enable native build support for DS using the same data migration pipline with a single native executable.
```

====

> Also, for those who are on board on Kubernetes, you can run the same pipeline through the https://debezium.io/documentation/reference/stable/operations/debezium-operator.html[Debezium Operator].

Reviewer 3: ```suggestion
If you are using Kubernetes, the same pipeline can be deployed using the https://debezium.io/documentation/reference/stable/operations/debezium-operator.html[Debezium Operator].
```

> After having the Operator installed on you cluster you can just create the following CRD

Reviewer 3: ```suggestion
After installing the Operator on your cluster, you can create the following Custom Resource Definition (CRD).
```

.DebeziumServer custom resource
[source, yaml]
----
  apiVersion: debezium.io/v1alpha1
  kind: DebeziumServer
  metadata:
    name: jdbc-sink-example
    namespace: debezium
  spec:
    image: quay.io/debezium/server:latest
    sink:
      type: jdbc
      config:
        connection.url: jdbc:mysql://mysql:3306/target_db
        connection.username: root
        connection.password: debezium
        insert.mode: upsert
        primary.key.mode: record_key
        schema.evolution: basic
        delete.enabled: "true"
    source:
      class: io.debezium.connector.postgresql.PostgresConnector
      config:
        database.hostname: postgres
        database.port: "5432"
        database.user: postgres
        database.password: postgres
        database.dbname: postgres
        topic.prefix: migration
        schema.include.list: inventory
        table.include.list: inventory.customers,inventory.products,inventory.products_on_hand,inventory.orders
    quarkus:
      config:
        log.console.json: "false"
----

== Data Integration with Debezium Management Platform

> Debezium Platform lets you manage your pipeline with a modern UI focusing on its design without worry about the deployment.

Reviewer 3: ```suggestion
Debezium Platform lets you manage your pipeline with a modern UI, while Debezium orchestrates the entire deployment automatically.
```

> Once you have https://debezium.io/documentation/reference/stable/operations/debezium-platform.html#_installation[installed] it on you Kubernetes cluster in `debezium-platform` namespace, you can create your data replication pipeline with easy

Reviewer 3: ```suggestion
Once the platform is https://debezium.io/documentation/reference/stable/operations/debezium-platform.html#_installation[installed] on your Kubernetes cluster in the `debezium-platform` namespace, you can create your data replication pipeline with ease.
```

> Once you have https://debezium.io/documentation/reference/stable/operations/debezium-platform.html#_installation[installed] it on you Kubernetes cluster in `debezium-platform` namespace, you can create your data replication pipeline with easy

Reviewer 4: ```suggestion
Once you have https://debezium.io/documentation/reference/stable/operations/debezium-platform.html#_installation[installed] it on you Kubernetes cluster in `debezium-platform` namespace, you can create your data replication pipeline with easy.
```

[NOTE]
====
Through the https://summerofcode.withgoogle.com/[Google Summer of code initiative] we have a contributor working to support host based deploy on the Debezium Platform.

> This means that in future you could deploy pipeline also on VMs, on-prem machine and cloud.

Reviewer 3: ```suggestion
This means that in the future, you could also deploy the pipeline on VMs, on-prem machines, and the cloud.
```

> This means that in future you could deploy pipeline also on VMs, on-prem machine and cloud.

Reviewer 4: ```suggestion
This means that in the future you could deploy pipeline also on the VMs, on-prem machine and cloud.
```

Keep an eye on https://github.com/debezium/dbz/issues/2085
====

> You can create the following resources to create the source and destination databases

Reviewer 3: ```suggestion
You can use the following resources to create the source and destination databases.
Once applied, a PostgreSQL database acting as the source will be available as `postgresql`, and another PostgreSQL database acting as the destination will be available as `postgresql-dest`.
```

</div>
++++

Then create the source and the destination

*Create the source*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-source.png" style="max-width:100%;" class="responsive-image" alt="Create the source">
</div>
++++

*Create the destination*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-destination.png" style="max-width:100%;" class="responsive-image" alt="Create the destination">
</div>
++++

Once you have all the pieces ready, you can design the pipeline that will replicate data from the source to the destination database

*Design the pipeline*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/design-pipeline.png" style="max-width:100%;" class="responsive-image" alt="Pipeline design">
</div>
++++

and the finally create it

*Create the pipeline*

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-06-22-kafka-less-migration/create-pipeline.png" style="max-width:100%;" class="responsive-image" alt="Create pipeline">
</div>
++++

> It is worth to mention that you can do the same thing via the REST API with the following

Reviewer 3: ```suggestion
It is worth mentioning that you can do the same thing via the REST API with the following
```

"password": "debezium"
}
}'
----

.Create source (use the connection id from step 1):
[source, shell]
----
curl -X POST http://platform.debezium.io/api/sources \
-H 'Content-Type: application/json' \
-d '{
"name": "pg-source",
"type": "io.debezium.connector.postgresql.PostgresConnector",
"schema": "dummy",
"connection": { "id": 1 },
"config": {
"schema.include.list": "inventory",
"table.exclude.list": "inventory.geom,inventory.spatial_ref_sys",
"topic.prefix": "inventory"
}
}'
----

.Create destination (use the connection id from step 2):
[source, shell]
----
curl -X POST http://platform.debezium.io/api/destinations \
-H 'Content-Type: application/json' \
-d '{
"name": "jdbc-destination",
"type": "jdbc",
"schema": "dummy",
"connection": { "id": 2 },
"config": {
"insert.mode": "upsert",
"primary.key.mode": "record_key",
"schema.evolution": "basic"
}
}'
----

.Create pipeline (use the source and destination ids from steps 3 and 4):
[source, shell]
----
curl -X POST http://platform.debezium.io/api/pipelines \
-H 'Content-Type: application/json' \
-d '{
"name": "database-migration",
"source": { "id": 1 },
"destination": { "id": 1 },
"logLevel": "INFO"
}'
----

== Conclusion

> We have seen how simple it is to build a data migration pipeline without requiring a Kafka cluster.

Reviewer 3: ```suggestion
We have seen how simple it is to build a data migration pipeline without requiring Kafka.
```

There are still some limitations in the current DS JDBC sink implementation.

> For example, while Kafka Connect (KC) allows processing to be parallelized by configuring `tasks.max`, this capability is not yet available in DS.

Reviewer 3: ```suggestion
For example, while Kafka Connect (KC) allows processing to be parallelized by configuring `tasks.max` with a value greater than `1`, this capability is not yet available in DS.
```

Similarly, if you want to process records from only a specific set of tables, there is no direct equivalent to the KC topics configuration.

> Instead, you need to filter records at the source, using the `table.include.list` configuration, or apply a transformation to select the desired events.

Reviewer 3: I am not sure I would necessarily frame this as a limitation. One could argue that the same semantics apply to every DS consumer, and given that it's P2P, I think it's implied that this is controlled more at the source than at the target.

I think instead this point makes more sense to be framed that in Kafka, producers and consumers are decoupled, which provides a certain amount of latitude in how a consumer can ingest changes selectively, whereas in DS, where the consumer is directly associated with the producer, this responsibility shifts to the producer because there is no durable middleware in between.

So rather than a limitation, it's an intentional design choice.

WDYT?

Author: Yes, sounds good. I re worked it. Let me know WDYT now, @Naros

These are known limitations, and addressing them is already on our radar for future releases.

If you have specific requirements or feature requests for Debezium Server or the Debezium Management Platform, we'd love to hear from you.

> Your feedback helps us prioritize what matters most to the community, so feel free to reach out via the https://groups.google.com/g/debezium[mailing list] or https://debezium.zulipchat.com/[Zulip chat].

Reviewer 4: Maybe encourage the user to create directly the feature requests or bug reports under https://github.com/debezium/dbz/issues
